ksqldb-parser/src/main/java/io/confluent/ksql/parser/VariableSubstitutor.java

```diff
@@ -0,0 +1,212 @@
+      this.valueMap = requireNonNull(valueMap, "valueMap");
+      this.sanitizedValueMap = new HashMap<>(valueMap.size());
+    }
+
+    public String replace(final SqlBaseParser.SingleStatementContext singleStatementContext) {
+      // walk the statement tree to validate and sanitize variables
+      visit(singleStatementContext);
+
+      // replace all variables with sanitized values
+      return StringSubstitutor.replace(statementText, sanitizedValueMap);
+    }
+
+    @Override
+    public Void visitStringLiteral(final SqlBaseParser.StringLiteralContext context) {
+      final String text = unquote(context.getText(), "\'");
+      for (String variableName : VariablesLookup.lookup(text)) {
+        if (valueMap.containsKey(variableName)) {
+          sanitizedValueMap.putIfAbsent(variableName, sanitize(valueMap.get(variableName)));
+        }
+      }
+
+      return null;
+    }
+
+    @Override
+    public Void visitVariableLiteral(final SqlBaseParser.VariableLiteralContext context) {
+      final String variableRef = context.getText();
+      final String variableName = unwrap(variableRef);
+      final String variableValue = valueMap.getOrDefault(variableName, variableRef);
+
+      throwIfInvalidLiteral(variableValue, getLocation(context));
+      sanitizedValueMap.putIfAbsent(variableName, sanitize(variableValue));
+      return null;
+    }
+
+    @Override
+    public Void visitVariableIdentifier(final SqlBaseParser.VariableIdentifierContext context) {
+      final String variableRef = context.getText();
+      final String variableName = unwrap(variableRef);
+      final String variableValue = valueMap.getOrDefault(variableName, variableRef);
+
+      throwIfInvalidIdentifier(variableValue, getLocation(context));
+      sanitizedValueMap.putIfAbsent(variableName, variableValue);
+      return null;
+    }
+
+    private String getIdentifierText(final String value) {
+      final char firstChar = value.charAt(0);
+      final char lastChar = value.charAt(value.length() - 1);
+
+      if (firstChar == '"' && lastChar == '"') {
+        return unquote(value, "\"");
+      } else if (firstChar == '`' && lastChar == '`') {
+        return unquote(value, "`");
+      } else {
+        return value;
+      }
+    }
+
+    static String unwrap(final String value) {
+      return value.substring(PREFIX.length(), value.length() - SUFFIX.length());
+    }
+
+    private void throwIfInvalidIdentifier(
+        final String value,
+        final Optional<NodeLocation> location
+    ) {
+      final String identifierText = getIdentifierText(value);
+
+      if (!VALID_IDENTIFIER_NAMES.matcher(identifierText).matches()) {
+        throw new ParseFailedException(
+            "Illegal argument at " + location.map(NodeLocation::toString).orElse("?")
+                + ". Identifier names may only contain alphanumeric values, '_' "
+                + "or not starting with '@'. Got: '" + value + "'",
```

Comment:
The text is a bit confusing. You mean: "Identifier names cannot start with '@' and may only contain alphanumeric values and '_'. "

Reply:
Thanks. That sounds much better.
Done